quinn-proto/src/endpoint.rs

```diff
@@ -155,36 +168,32 @@ pub struct Context {
     pub incoming_handshakes: usize,
     pub dirty_conns: FnvHashSet<ConnectionHandle>,
     pub readable_conns: FnvHashSet<ConnectionHandle>,
-    pub listen_keys: Option<ListenKeys>,
 }
 
 /// Information that should be preserved between restarts for server endpoints.
 ///
 /// Keeping this around allows better behavior by clients that communicated with a previous
 /// instance of the same endpoint.
 pub struct ListenKeys {
```

Comment:
I think the design here ends up a little weird. We have `Config::use_stateless_retry` which switches whether we require stateless retries or not and we have `ListenKeys` which is required to be `Some(key)` if you want your endpoint to act as a server but whose key may go unused if `use_stateless_retry` is `false`. Maybe make `stateless_retry_key` an `Option<>` in `Config` which handles both in one go?

Reply:
And rename `ListenKey` to `StatelessRetryKey` in the process?

Reply:
If we want to support emitting `NEW_TOKEN` frames, then doing so could reasonably use this key whether or not stateless retries are enabled. Do you have any insight as to how likely we are to want that? If we don't think we'll want it for the forseeable future, I'm inclined to remove `Config::use_stateless_retry` and make the `ListenKeys` field `stateless_retry_key: Option<StatelessRetryKey>`, since its presence is only meaningful to a server.

Reply:
Maybe we should rename `ListenKeys` to `ListenConfig` or `ServerConfig` and move other server-specific members of `Config` (e.g. `accept_buffer`, `tls_server_config`, `retry_token_lifetime`) there?

Reply:
Yeah, doing a `ServerConfig` sounds like the right direction, and we can then use further `Option`s inside that to enable various features.